### PR TITLE
Feature dns templates

### LINKED ISSUE #8: Implement DNS Templates
Documentation: https://dm.realtimeregister.com/docs/api/templates

**Module:**

- [ ] Add a `TemplatesApi` module in the `Api` namespace
- [ ] Add this module to the `RealtimeRegister` class.

**Endpoints:**

- [ ] Get
- [ ] List
- [ ] Create
- [ ] Update
- [ ] Delete

When implementing the endpoints, make sure to include links to the documentation using the `@see` annotation in docblocks.

**Tests:**

- [ ] Add tests for all the added domain objects
- [ ] Add tests for all the added endpoints

### PR DESCRIPTION
**Pull request by jeroenvermeulen**
Implemented DnsTemplates API, fixes #8 

Example code to use:

```php
<?php
declare(strict_types=1);

require_once __DIR__ . "/vendor/autoload.php";

use SandwaveIo\RealtimeRegister\RealtimeRegister;
use SandwaveIo\RealtimeRegister\Domain\Zone;
use SandwaveIo\RealtimeRegister\Domain\DomainZoneRecordCollection;
use SandwaveIo\RealtimeRegister\Exceptions\NotFoundException;

$realtimeRegister = new RealtimeRegister('my-secret-api-key');

$customer = 'acmeinc';
$domain = 'dit-is-een-test.nl';
$templateName = 'redirect_demo';

try {
    $realtimeRegister->dnstemplates->get($customer, $templateName);
} catch (NotFoundException $e) {
    $realtimeRegister->dnstemplates->create(
        $customer,
        $templateName,
        'dns@acmeinc.com',
        86400,
        10800,
        3600000,
        3600,
        DomainZoneRecordCollection::fromArray(
            [
                [
                    'name'    => '##DOMAIN##',
                    'type'    => 'URL',
                    'content' => 'https://www.savvii.com/',
                    'ttl'     => 300
                ],
                [
                    'name'    => 'www.##DOMAIN##',
                    'type'    => 'URL',
                    'content' => 'https://www.savvii.com/',
                    'ttl'     => 300
                ]
            ]
        )
    );
}

$realtimeRegister->domains->update(
    $domain,
    zone: Zone::fromArray(['template' => $templateName])
);
```

### CODE REVIEW

**Comment by jeroenvermeulen** [comment]
@A1rPun  Anybody home?

File: src/Api/DnsTemplatesApi.php [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+            'retry'      => $retry,
+            'expire'     => $expire,
+            'ttl'        => $ttl
+        ];
+        if ($records) {
+            $payload['records'] = $records->toArray();
+        }
+
+        $this->client->post(
+            sprintf("v2/customers/%s/dnstemplates/%s/update", $customer, $name),
```
[the remark this answers is not in the document]
Comment (jeroenvermeulen) [reply src/Api/DnsTemplatesApi.php:151]: I agree this is better, so I will change. But you have a lot of places which also needs this change, check:
https://github.com/sandwave-io/realtimeregister-php/search?q=this-%3Eclient-%3Epost
and
https://github.com/sandwave-io/realtimeregister-php/search?q=this-%3Eclient-%3Eget

File: tests/Clients/DnsTemplatesApiListTest.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                    'total'  => 3,
+                    'offset' => 0,
+                    'limit'  => 3,
+                ],
+            ]),
+            MockedClientFactory::assertRoute('GET', 'v2/customers/johndoe/dnstemplates', $this)
+        );
+
+        $response = $sdk->dnstemplates->list('johndoe', 3, 0, 'john');
+        $this->assertInstanceOf(DnsTemplateCollection::class, $response);
```
[the remark this answers is not in the document]
Comment (jeroenvermeulen) [reply tests/Clients/DnsTemplatesApiListTest.php]: It does perform assertions, so I added an assert on the response count.

File: tests/Clients/DnsTemplatesApiListTest.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            MockedClientFactory::assertRoute('GET', 'v2/customers/johndoe/dnstemplates', $this, [
+                'name' => 'test',
+                'limit' => '3',
+                'offset' => '0',
+                'q' => 'john',
+            ])
+        );
+
+        $response = $sdk->dnstemplates->list('johndoe', 3, 0, 'john', $parameters);
+        $this->assertInstanceOf(DnsTemplateCollection::class, $response);
```
[the remark this answers is not in the document]
Comment (jeroenvermeulen) [reply tests/Clients/DnsTemplatesApiListTest.php]: It does perform assertions, so I added an assert on the response count.

**Comment by jeroenvermeulen** [comment]
I have made all requested changes

**Review by A1rPun — DISMISSED** [review]
Thanks for this PR

**Comment by jeroenvermeulen** [comment]
@arnowest Looks like phpcs **does** want the trailing comma's to be there:
https://github.com/sandwave-io/realtimeregister-php/actions/runs/4356273605/jobs/7615193232

**Comment by arnowest** [comment]
> @arnowest Looks like phpcs **does** want the trailing comma's to be there: https://github.com/sandwave-io/realtimeregister-php/actions/runs/4356273605/jobs/7615193232

The CS fixer is complaining about trailing comma's in arrays. Looking at the job output there is no mention of the trailing comma's in the function parameters (/constructor). If you let CS fixer fix all the reported issues this MR can be merged.

**Review by arnowest — DISMISSED** [review]

File: src/Api/DnsTemplatesApi.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (arnowest) [reply src/Api/DnsTemplatesApi.php]: Fair point! I've created an issue for resolving this:
https://github.com/sandwave-io/realtimeregister-php/issues/70

**Comment by jeroenvermeulen** [comment]
@arnowest OK, done

**Review by A1rPun — APPROVED** [review]

**Review by arnowest — APPROVED** [review]

**Pull request merged**